### ISSUE TITLE
[FSDP2] need dummy forward/backward to stay SPMD

### ISSUE THREAD

**Issue by weifengpy** [pytorch/pytorch#152964]
### 🚀 The feature, motivation and pitch

FSDP2 assumes SPMD on every rank, meaning every rank needs to call forward/backward to issue all-gather / reduce-scatter

However, user reported two cases that some rank might be skipping forward/backward
* torchtune might mask all the activations. they have to create a dummy input to avoid job hanging https://github.com/pbontrager/torchtune/blob/ee2f5999976ce476c8dcb8e0f09dc7258a9b704c/torchtune/modules/loss/cross_entropy_loss.py#L76
* dataloader might return empty input for certain ranks. TODO: add github issue

**Comment by fegin** [comment]
Do we want to do this in fully_shard or we would like users to ensure the inputs are SPMD-friendly?

**Comment by awgu** [comment]
Fixing this in `fully_shard` incurs extra complexity while obfuscating the contract. I would recommend requiring users to make their programs SPMD.

**Comment by weifengpy** [comment]
get a motivating example from HF transformers/models/mllama/modeling_mllama.py

https://github.com/huggingface/transformers/blob/a52478253bbe522a420e88ea3940d4d98a935300/src/transformers/models/mllama/modeling_mllama.py#L1712-L1723 

```
if pixel_values is not None:
    vision_outputs = self.vision_model(...)
outputs = self.language_model(...)
```

if we wrap vision_model with fully_shard, sometimes it's freezing. by nature, some training example (user id) does not have image

@awgu is the example making sense to you?
